importlib_metadata/__init__.py

```diff
@@ -249,72 +292,116 @@ class DeprecatedList(list):
         stacklevel=pypy_partial(2),
     )
 
-    def __setitem__(self, *args, **kwargs):
+    @overload
+    def __setitem__(self, index: SupportsIndex, value: _T) -> None:
+        ...  # pragma: no cover
+
+    @overload
+    def __setitem__(self, index: slice, value: Iterable[_T]) -> None:
+        ...  # pragma: no cover
+
+    def __setitem__(self, index: Union[SupportsIndex, slice], value: Any) -> None:
         self._warn()
-        return super().__setitem__(*args, **kwargs)
+        super().__setitem__(index, value)
 
-    def __delitem__(self, *args, **kwargs):
+    def __delitem__(self, index: Union[SupportsIndex, slice]) -> None:
         self._warn()
-        return super().__delitem__(*args, **kwargs)
+        super().__delitem__(index)
 
-    def append(self, *args, **kwargs):
+    def append(self, value: _T) -> None:
         self._warn()
-        return super().append(*args, **kwargs)
+        super().append(value)
 
-    def reverse(self, *args, **kwargs):
+    def reverse(self) -> None:
         self._warn()
-        return super().reverse(*args, **kwargs)
+        super().reverse()
 
-    def extend(self, *args, **kwargs):
+    def extend(self, values: Iterable[_T]) -> None:
         self._warn()
-        return super().extend(*args, **kwargs)
+        super().extend(values)
 
-    def pop(self, *args, **kwargs):
+    def pop(self, index: int = -1) -> _T:
```

Comment:
Yes. I like that, and it removes copy/paste code. One thing to consider/check is that the stacklevel is still correct in the warning, but I think it would be. I'd like to see this refactor adding prior to adding the annotations. I can do that.

Reply:
I tried implementing it as written, but when I did so, the `super()` call failed:

```
269     >>> dl[0] = 1
UNEXPECTED EXCEPTION: RuntimeError('super(): __class__ cell not found')
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/doctest.py", line 1348, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest importlib_metadata.DeprecatedList[2]>", line 1, in <module>
  File "/Users/jaraco/code/public/importlib_metadata/importlib_metadata/__init__.py", line 257, in wrapped
    return getattr(super(), method_name)(*args, **kwargs)
RuntimeError: super(): __class__ cell not found
/Users/jaraco/code/public/importlib_metadata/importlib_metadata/__init__.py:269: UnexpectedException
```

Moving the `_wrap_deprecated_method` into the class namespace solved that issue.

Reply:
As I was working on an implementation, I ran into an emergent failure. It seems that in the past couple of hours, the tests have [started failing on Python 3.10](https://github.com/python/importlib_metadata/runs/3455554997?check_suite_focus=true). That's a re-run of the test on main that previously passed, the same code as was released with the v4.8.1 tag. I'll need to track that down first.

Reply:
In 5875f39205e, I implement the alternative. Annoyingly, I had to apply `type: ignore` to the `_wrap_deprecated_method` because mypy did not like a function without a `self` parameter:

```
293: error: Self argument missing for a non-static method (or an invalid type for self)
```

Maybe you'll have some ideas on how to improve the type declaration around that usage.